propagation.go

```diff
@@ -289,12 +314,191 @@ func (p *textMapPropagator) parseCommaSeparatedMap(value string) map[string]stri
 }
+
+// CompositePropagator can be used to extract and inject OpenTracing headers into SpanContexts using multiple
+// propagators.
+type CompositePropagator struct {
+	propagators []Propagator
+}
+
+// NewCompositePropagator creates a Propagator that can be used to perform extraction and injection of SpanContexts from
+// and into carriers from multiple Propagators.
+func NewCompositePropagator(propagators []Propagator) *CompositePropagator {
+	return &CompositePropagator{
+		propagators: propagators,
+	}
+}
+
+// Inject attempts to inject a SpanContext into a given carrier using the underlying propagators. If one of the
+// propagators is unable to inject the SpanContext, this method errors out (causing injection attempts using the
+// subsequent propagators to be skipped).
+func (p *CompositePropagator) Inject(sc SpanContext, abstractCarrier interface{}) error {
+	for _, propagator := range p.propagators {
+		err := propagator.Inject(sc, abstractCarrier)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Extract attempts to extract a SpanContext from the given carrier using the underlying propagators.
+//
+// If multiple propagators are capable of extracting a valid context, this propagator will ensure the contexts match
+// eachother. Contexts are considered as non-clashing as long as their IDs and baggage match (or extend) eachother. If
+// the extracted contexts clash (e.g. by having different values for what would be the same baggage item or different
+// IDs), an ErrSpanContextCorrupted will be returned.
+//
+// If no underlying propagator is able to extract a valid span, all the errors will be combined together in order to
+// produce a single (opentracing-compliant) error (see mergeExtractedError for specifics on this).
+func (p *CompositePropagator) Extract(abstractCarrier interface{}) (SpanContext, error) {
+	extractedContexts := make([]SpanContext, 0)
+	extractedErrors := make([]error, 0)
+
+	for _, propagator := range p.propagators {
+		sc, err := propagator.Extract(abstractCarrier)
+
+		if err != nil {
+			extractedErrors = append(extractedErrors, err)
+		} else {
+			extractedContexts = append(extractedContexts, sc)
+		}
+	}
+
+	return combineExtractResults(extractedContexts, extractedErrors)
+}
+
+func combineExtractResults(contexts []SpanContext, errors []error) (SpanContext, error) {
+	ret := emptyContext
+	var err error
+
+	for _, sc := range contexts {
+		// Base case
+		if !ret.IsValid() && !ret.isDebugIDContainerOnly() {
+			ret = sc
+		} else if sc.IsValid() {
+			// If the context is valid (i.e. a full span context), it should either be merged with another valid context
+			// (that we've already extracted) or a debug container
+			if ret.IsValid() {
+				ret, err = mergeValidContexts(ret, sc)
+			} else {
+				ret, err = mergeContextWithDebugIDContainer(sc, ret)
+			}
+
+			if err != nil {
+				return emptyContext, opentracing.ErrSpanContextCorrupted
+			}
+		} else if sc.isDebugIDContainerOnly() {
```

Comment:
as I mentioned earlier, I think dealing with debug-id/jaeger-baggage everywhere is complicating the code and lot. All the custom codecs should be concerned with debug-id/jaeger-baggage, it should only be done by a single class.